apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,483 @@
+
+  /**
+   * Hold instance of A2DP status changed handler.
+   *
+   * @private
+   */
+  _bindA2dpStatusChangedHandler: null,
+
+  /**
+   * Hold instance of SCO status changed handler.
+   *
+   * @private
+   */
+  _bindScoStatusChangedHandler: null,
+
+  /**
+   * Store a reference for mozBluetooth.
+   *
+   * @private
+   */
+  _bluetooth: null,
+
+  /**
+   * State of Bluetooth default adapter.
+   *
+   * @private
+   */
+  _isEnabled: false,
+
+  /**
+   * Build-in Bluetooth profiles.
+   *
+   * @public
+   */
+  get Profiles() {
+    return {
+      HFP: 'hfp',   // Hands-Free Profile
+      OPP: 'opp',   // Object Push Profile
+      A2DP: 'a2dp', // A2DP status
+      SCO: 'sco'    // Synchronous Connection-Oriented
+    };
+  },
+
+  /**
+   * Set profile connect state.
+   *
+   * @private
+   * @param {String} profile profile id
+   * @param {Boolean} connected connect status
+   */
+  _setProfileConnected: function bt_setProfileConnected(profile, connected) {
+    var value = this['_' + profile + 'Connected'];
+    if (value !== connected) {
+      this['_' + profile + 'Connected'] = connected;
+
+      // Raise an event for the profile connection changes.
+      window.dispatchEvent(new CustomEvent('bluetoothprofileconnectionchange',
+      {
+        name: profile,
+        connected: connected
+      }));
+      if (profile === 'opp' && this.transferIcon) {
+        this.transferIcon.update();
+      }
+      if (profile === 'a2dp' && this.headphoneIcon) {
+        this.headphoneIcon.update();
+      }
+    }
+  },
+
+  /**
+   * Check if bluetooth profile is connected.
+   *
+   * @public
+   * @param {String} profile profile name
+   * @return {Boolean} connected state
+   */
+  isProfileConnected: function bt_isProfileConnected(profile) {
+    var isConnected = this['_' + profile + 'Connected'];
+    if (isConnected === undefined) {
+      return false;
+    } else {
+      return isConnected;
+    }
+  },
+
+  /**
+   * Initialize bluetooth module.
+   *
+   * @public
+   */
+  start: function bt_start() {
+    if (!window.navigator.mozSettings || !window.navigator.mozBluetooth) {
+      return;
+    }
+
+    this._bluetooth = window.navigator.mozBluetooth;
+
+    // hold the instances so we can remove it when needed
+    this._bindBtAdapterHandler =
+      this._btAdapterHandler.bind(this);
+
+    this._bindHfpStatusChangedHandler = function(evt) {
+      this._setProfileConnected(this.Profiles.HFP, evt.status);
+    }.bind(this);
+    this._bindA2dpStatusChangedHandler = function(evt) {
+      this._setProfileConnected(this.Profiles.A2DP, evt.status);
+    }.bind(this);
+    this._bindScoStatusChangedHandler = function(evt) {
+      this._setProfileConnected(this.Profiles.SCO, evt.status);
+    }.bind(this);
+
+    // listen states when the adapter is available
+    window.addEventListener('bluetooth-available',
+      this._adapterAvailableHandler.bind(this));
+
+    this._initDefaultAdapter();
```

Comment:
This `_initDefaultAdapter ` call should be moved after `addEventListener('attributechanged',`(Line 176). Because we might miss to catch attribute change event, if we access `this._adapter = this._bluetooth.defaultAdapter;` (Line 218) in `else` case. Then, platform will fire attribute changed event immediately. This is a race condition potentially.